### ISSUE TITLE
Assumption wrapping bug + additional wrapping issues

### ISSUE THREAD

**Issue by cheshire** [sosy-lab/java-smt#104]
The method `addConstraint` does not remove the assumptions, and as a result the following code produces the wrong result:

```java
    BooleanFormula a = bmgr.makeVariable("a");
    BooleanFormula b = bmgr.makeVariable("b");
    try (ProverEnvironment pe = context.newProverEnvironment()) {
      pe.push();
      assertThat(pe.isUnsatWithAssumptions(ImmutableSet.of(bmgr.not(a)))).isFalse();
      pe.addConstraint(a);
      assertThat(pe.isUnsatWithAssumptions(ImmutableSet.of(bmgr.not(a)))).isTrue();
```

The bug is that clearing assumptions mistakenly kills the user-defined constraint.

I think the code quality of the wrapping feature can be substantially improved: now, it is not obvious whether `clearAssumptions` is properly called everywhere without examining 3-4 classes inheriting from each other. Ideally, clearing should be atomic, and it should be obvious by looking at a single method.

Additionally, I don't even quite understand the purpose of the wrapper: in my understanding, the purpose of the assumptions feature is to be able to get an UNSAT core over them. Yet the wrapper obviously can not do that, as we are not implementing our own solver.
So to me it seems to be even worse than just throwing the exception on the `isUnsatWithAssumptions` call, as a subsequent `unsatCoreOverAssumptions` may and will return the wrong answer.

I am even further confused by the integration of the wrapper with interpolation, especially by a visitor removing assumed variables. From my understanding, they should never be an interpolant anyway, if they are asserted only on one side on the interpolation query, which follows from the definition.

@kfriedberger is there any benefit in keeping the wrapper?

**Comment by kfriedberger** [comment]
I have not implemented the class `ProverWithAssumptionsWrapper`, but whoever has done this might have had his reasons. If there is no use-case, e.g., in CPAchecker, then we might think about removing it.

**Comment by cheshire** [comment]
Right, maybe it was @stieglma ?.. In any case, haven't found any usages in CPAchecker apart from my weakening code, and the handling is incorrect in any case, as we don't provide an unsat core over assumptions. Thus I would like to remove this.

**Comment by kfriedberger** [comment]
partially fixed in 48757287ad622bd91f60dc8a52fdf11db0268e76.
I like the idea of the wrapper.
Perhaps we need to discuss the behaviour of `unsatCoreOverAssumptions` and implement it in the right way.

**Comment by kfriedberger** [comment]
The bug mentioned in this issue was fixed some time ago.
In my opinion, this issue can be closed for now.
